### PR TITLE
Add deployment code for postgres 

### PR DESCRIPTION
**Pull request by bj2k61**
## Trello card URL  https://trello.com/c/JSsTue0b/492-deploy-postgres

How to review  
cd into teaching-vacancies  make review_aks terraform-init-aks pr_id=postgres VERBOSE=1
make review_aks terraform-plan-aks pr_id=postgres VERBOSE=1

### CODE REVIEW

File: terraform/aks/variables.tf [hunk 1]
```diff
@@ -0,0 +1,58 @@
...
+variable "deploy_azure_backing_services" {
+  default     = true
+  description = "Deploy real Azure backing services like databases, as opposed to containers inside of AKS"
+}
+variable "enable_postgres_ssl" {
+  default     = true
+  description = "Enforce SSL connection from the client side"
+}
+variable "docker_image" {
+  description = "Docker image full name to identify it in the registry. Includes docker registry, repository and tag e.g.: ghcr.io/dfe-digital/teacher-pay-calculator:673f6309fd0c907014f44d6732496ecd92a2bcd0"
```
Comment (smneal) [inline terraform/aks/variables.tf:35]: Could this be updated for the local project?

File: terraform/aks/vendor/modules/aks
Comment (smneal) [inline terraform/aks/vendor/modules/aks:1]: I think we usually exclude the vendor folder from check-ins.

File: terraform/aks/database.tf [hunk 2]
```diff
@@ -0,0 +1,17 @@
...
+  azure_resource_prefix = var.azure_resource_prefix
+  service_name          = var.service_name
+  service_short         = var.service_short
+  config_short          = var.config_short
+
+  cluster_configuration_map = module.cluster_data.configuration_map
+
+  use_azure               = var.deploy_azure_backing_services
+  azure_enable_monitoring = false
+  server_version          = "14"
```
Comment (smneal) [inline terraform/aks/database.tf:15]: The default version of the module is 14, so this could be removed.

File: terraform/aks/database.tf [hunk 3]
```diff
@@ -0,0 +1,17 @@
...
+  environment           = var.environment
+  azure_resource_prefix = var.azure_resource_prefix
+  service_name          = var.service_name
+  service_short         = var.service_short
+  config_short          = var.config_short
+
+  cluster_configuration_map = module.cluster_data.configuration_map
+
+  use_azure               = var.deploy_azure_backing_services
+  azure_enable_monitoring = false
```
Comment (smneal) [inline terraform/aks/database.tf:14]: We could set this from a variable in the tfvars.json file.

File: bin/terrafile
Comment (smneal) [inline bin/terrafile:1]: We normally exclude this from check-ins too.

File: Makefile.bak1
Comment (smneal) [inline Makefile.bak1:1]: This needs removing.

**Comment by smneal** [comment]
`terraform/aks/database.tf` needs formatting so it passes the `terraform fmt` test

File: terraform/aks/vendor/modules/aks
Comment (bj2k61) [reply terraform/aks/vendor/modules/aks:1]: i have removed this folder

File: Makefile.bak1
Comment (bj2k61) [reply Makefile.bak1:1]: removed

File: bin/terrafile
Comment (bj2k61) [reply bin/terrafile:1]: this has been removed

File: terraform/aks/database.tf (hunk 2 above)
Comment (bj2k61) [reply terraform/aks/database.tf]: this has been removed

File: terraform/aks/database.tf (hunk 3 above)
Comment (bj2k61) [reply terraform/aks/database.tf]: this has been moved as recommended

File: Makefile [hunk 4]
```diff
@@ -51,6 +51,49 @@ dev: ## dev
 		$(eval var_file=$(env))
 		$(eval backend_config=-backend-config="key=$(env)/app.tfstate")
 
+
+.PHONY: development_aks
+development_aks:
```
Comment (saliceti) [inline Makefile:56]: We are focusing on review, this can be removed for now

File: Makefile [hunk 5]
```diff
@@ -51,6 +51,49 @@ dev: ## dev
...
+	$(eval include global_config/development.sh)
+
+
+.PHONY: set-key-vault-names
+set-key-vault-names:
```
Comment (saliceti) [inline Makefile:61]: We won't use key vault so this can be removed

File: Makefile [hunk 6]
```diff
@@ -51,6 +51,49 @@ dev: ## dev
...
+		| tar xz -C ./bin terrafile
+
+terraform-init-aks: composed-variables bin/terrafile set-azure-account
+	$(if ${IMAGE_TAG}, , $(eval IMAGE_TAG=master))
+
+	./bin/terrafile -p terraform/aks/vendor/modules -f terraform/aks/config/$(CONFIG)_Terrafile
+	terraform -chdir=terraform/aks init -upgrade -reconfigure \
+		-backend-config=resource_group_name=${RESOURCE_GROUP_NAME} \
+		-backend-config=storage_account_name=${STORAGE_ACCOUNT_NAME} \
+		-backend-config=key=${ENVIRONMENT}_kubernetes.tfstate
```
Comment (saliceti) [inline Makefile:76]: This can simply be `${ENVIRONMENT}.tfstate`

File: Makefile [hunk 7]
```diff
@@ -51,6 +51,49 @@ dev: ## dev
...
+
+	$(eval export TF_VAR_azure_resource_prefix=${AZURE_RESOURCE_PREFIX})
+	$(eval export TF_VAR_config_short=${CONFIG_SHORT})
+	$(eval export TF_VAR_service_name=${SERVICE_NAME})
+	$(eval export TF_VAR_service_short=${SERVICE_SHORT})
+	$(eval export TF_VAR_docker_image=${DOCKER_REPOSITORY}:${IMAGE_TAG})
```
Comment (saliceti) [inline Makefile:82]: We don't need the docker image right now

File: Makefile [hunk 8]
```diff
@@ -51,6 +51,49 @@ dev: ## dev
...
+
+terraform-plan-aks: terraform-init-aks
+	terraform -chdir=terraform/aks plan -var-file "config/${CONFIG}.tfvars.json"
+
```
Comment (saliceti) [inline Makefile:86]: Please keep only 1 blank line

File: terraform/aks/config/development.tfvars.json
Comment (saliceti) [inline terraform/aks/config/development.tfvars.json:1]: We are focusing on review, this file can be removed

File: terraform/aks/config/development_Terrafile
Comment (saliceti) [inline terraform/aks/config/development_Terrafile:1]: We are focusing on review, this file can be removed

File: terraform/aks/database.tf [hunk 9]
```diff
@@ -0,0 +1,13 @@
...
+
+  namespace             = var.namespace
+  environment           = var.environment
+  azure_resource_prefix = var.azure_resource_prefix
+  service_name          = var.service_name
+  service_short         = var.service_short
+  config_short          = var.config_short
+  cluster_configuration_map = module.cluster_data.configuration_map
+  use_azure               = var.deploy_azure_backing_services
+  azure_enable_backup_storage  = var.azure_enable_backup_storage
```
Comment (saliceti) [inline terraform/aks/database.tf:12]: Please align (use format document in vscode or run terraform fmt)

File: terraform/aks/terraform.tf [hunk 10]
```diff
@@ -0,0 +1,41 @@
...
+}
+
+provider "kubernetes" {
+  host                   = module.cluster_data.kubernetes_host
+  client_certificate     = module.cluster_data.kubernetes_client_certificate
+  client_key             = module.cluster_data.kubernetes_client_key
+  cluster_ca_certificate = module.cluster_data.kubernetes_cluster_ca_certificate
+}
+
+provider "statuscake" {
```
Comment (saliceti) [inline terraform/aks/terraform.tf:39]: This is not required right now

File: terraform/aks/variables.tf (hunk 1 above)
Comment (saliceti) [reply terraform/aks/variables.tf:35]: Docker image is not required right now

File: terraform/aks/variables.tf [hunk 11]
```diff
@@ -0,0 +1,58 @@
...
+}
+variable "external_url" {
```
Comment (saliceti) [inline terraform/aks/variables.tf:37]: Please remove all unused variables

File: .gitignore
Comment (saliceti) [inline .gitignore:1]: This should have the vendor modules directory to exclude it from git

File: Makefile (hunk 4 above)
Comment (bj2k61) [reply Makefile:56]: This has been removed

File: Makefile (hunk 7 above)
Comment (bj2k61) [reply Makefile:82]: This has been removed

File: terraform/aks/config/development.tfvars.json
Comment (bj2k61) [reply terraform/aks/config/development.tfvars.json:1]: file has been removed

File: terraform/aks/config/development_Terrafile
Comment (bj2k61) [reply terraform/aks/config/development_Terrafile:1]: this has been removed

File: terraform/aks/database.tf (hunk 9 above)
Comment (bj2k61) [reply terraform/aks/database.tf:12]: done

File: terraform/aks/variables.tf (hunk 11 above)
Comment (bj2k61) [reply terraform/aks/variables.tf:37]: this has been resolved

File: .gitignore
Comment (bj2k61) [reply .gitignore:1]: directories has been added to gitignore

File: terraform/aks/terraform.tf (hunk 10 above)
Comment (bj2k61) [reply terraform/aks/terraform.tf:39]: this has been removed

File: Makefile (hunk 5 above)
Comment (bj2k61) [reply Makefile:55]: resolved

File: .gitignore [hunk 12]
```diff
@@ -96,3 +99,14 @@ dump.rdb
 
 # Visual Studio Code
 .vscode
+
+.idea
+*-export
```
Comment (saliceti) [inline .gitignore:104]: `*-export` and `/new_service` are specific to teacher-services-cloud repo, I'm not sure they're relevant here

File: terraform/aks/variables.tf (hunk 11 above)
Comment (saliceti) [reply terraform/aks/variables.tf:34]: I don't think external_url, statuscake_contact_groups are in use

File: .gitignore (hunk 12 above)
Comment (bj2k61) [reply .gitignore:104]: i have removed them

File: terraform/aks/variables.tf (hunk 11 above)
Comment (bj2k61) [reply terraform/aks/variables.tf:34]: both have been removed as suggested

File: terraform/aks/vendor/modules/aks
Comment (smneal) [inline terraform/aks/vendor/modules/aks:1]: You need to exclude the vendor folder so added it to the .gitignore file.

File: global_config/development.sh
Comment (smneal) [inline global_config/development.sh:1]: Like the tfvar files, this can be removed as the focus here is on review

File: .gitignore [hunk 13]
```diff
@@ -96,3 +100,10 @@ dump.rdb
 
 # Visual Studio Code
 .vscode
+.idea
+
+# devops files
+bin/terrafile
```
Comment (RMcVelia) [inline .gitignore:106]: duplicate entry

File: .gitignore [hunk 14]
```diff
@@ -96,3 +100,10 @@ dump.rdb
...
+
+# Downloaded terraform modules
+terraform/aks/vendor/
```
Comment (RMcVelia) [inline .gitignore:109]: duplicate entry, although the other one is terraform/aks/vendor without the ending / 
so just one or the other

**Comment by RMcVelia** [comment]
also please clean up commits

File: vendor/.keep
Comment (RMcVelia) [inline vendor/.keep:1]: why are we deleting this file?

File: global_config/review.sh [hunk 15]
```diff
@@ -1,4 +1,4 @@
-CONFIG=review_aks
+CONFIG=review
 var_file=${CONFIG}
```
Comment (nickdevdfe) [inline global_config/review.sh:2]: var_file needs to be hard set

Comment (bj2k61) [reply global_config/review.sh]: changed as recommended

**Pull request merged**